### ISSUE TITLE
false positive: np.array_split(): type of array_split() is partially unknown

### ISSUE THREAD

**Issue by austindonnelly** [microsoft/pyright#9819]
**Describe the bug**

Pylance in VS Code incorrectly flags calls to np.array_split() as partially unknown types.

**Code or Screenshots**

```python
import numpy as np

intlist: list[int] = [1, 2, 3]
bits = np.array_split(intlist, 2) # bug: Type of "array_split" is partially unknown flagged here.

assert len(bits) == 2
assert np.all(bits[0] == np.array([1,2]))
assert np.all(bits[1] == np.array([3]))
```

numpy==1.23.5

**VS Code extension or command-line**

Pyright is the version integrated into VS Code Pylance extension v2024.12.1
"Strict" mode is enabled via VS Code settings.json
``` json
    "python.analysis.typeCheckingMode": "strict",
```

**Analysis**

Numpy's shape_base.pyi stub declares two overloads for array_split():

``` python

@overload
def array_split(
    ary: _ArrayLike[_SCT],
    indices_or_sections: _ShapeLike,
    axis: SupportsIndex = ...,
) -> list[NDArray[_SCT]]: ...
@overload
def array_split(
    ary: ArrayLike,
    indices_or_sections: _ShapeLike,
    axis: SupportsIndex = ...,
) -> list[NDArray[Any]]: ...

```

The second overload is selected correctly, because the `intlist` is not a numpy array but a built-in `list`.
This should satisfy the ArrayLike option  `_NestedSequence[T]` where T=int.  That is completely known.

I would have expected Pylance to pick this second overload, realise the types are known, and infer a return type of `list[NDArray[Any]]`.
What actually happens is Plance flags this as:
```
Type of "array_split" is partially unknown
  Type of "array_split" is "Overload[(ary: _SupportsArray[dtype[_SCT@array_split]] | _NestedSequence[_SupportsArray[dtype[_SCT@array_split]]], indices_or_sections: SupportsIndex | Sequence[SupportsIndex], axis: SupportsIndex = ...) -> list[ndarray[Any, dtype[_SCT@array_split]]],
 (ary: _SupportsArray[dtype[Unknown]] | _NestedSequence[_SupportsArray[dtype[Unknown]]] | bool | int | float | complex | str | bytes | _NestedSequence[bool | int | float | complex | str | bytes], indices_or_sections: SupportsIndex | Sequence[SupportsIndex], axis: SupportsIndex = ...) -> list[ndarray[Any, dtype[Any]]]]"
Pylance[reportUnknownMemberType]
```

**Bonus**

The case I actually want to work is not an `intlist` but an object list. However, the numpy type stubs don't seem to support this, as they declare:

``` python
ArrayLike = _DualArrayLike[
    dtype,
    Union[bool, int, float, complex, str, bytes],
]
```

which would limit list[T] to only the T in that union, which doesn't include object.  However, that's a numpy bug, not pyright.

**Comment by erictraut** [comment]
You're using a very old version of numpy. In version 1.23.5 of the library, the annotations for the `array_split` function are incomplete. In particular, the `ary` parameter annotation is `ArrayLike`, which is a generic type alias that expects a single type argument, but none is provided in this case. That means the type argument defaults to "Unknown". This was fixed in later versions of numpy.

Your options are:
1. Upgrade to a newer version of numpy.
2. Do not enable "strict" mode type checking.
3. Suppress this particular error using a `# type: ignore` or `# pyright: ignore` comment.
